README.md
# peer-id

[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://protocol.ai)
[![](https://img.shields.io/badge/project-libp2p-yellow.svg?style=flat-square)](http://libp2p.io/)
[![](https://img.shields.io/badge/freenode-%23libp2p-yellow.svg?style=flat-square)](http://webchat.freenode.net/?channels=%23libp2p)
[![Discourse posts](https://img.shields.io/discourse/https/discuss.libp2p.io/posts.svg)](https://discuss.libp2p.io)
[![](https://img.shields.io/codecov/c/github/libp2p/js-peer-id.svg?style=flat-square)](https://codecov.io/gh/libp2p/js-peer-id)
[![](https://img.shields.io/travis/libp2p/js-peer-id.svg?style=flat-square)](https://travis-ci.com/libp2p/js-peer-id)
[![Dependency Status](https://david-dm.org/libp2p/js-peer-id.svg?style=flat-square)](https://david-dm.org/libp2p/js-peer-id)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/feross/standard)

> [IPFS](https://github.com/ipfs/ipfs) Peer ID implementation in JavaScript.

## Lead Maintainer

[Vasco Santos](https://github.com/vasco-santos)

## Table of Contents

- [peer-id](#peer-id)
  - [Lead Maintainer](#lead-maintainer)
  - [Table of Contents](#table-of-contents)
- [Description](#description)
- [Example](#example)
- [Installation](#installation)
  - [npm](#npm)
- [Setup](#setup)
  - [Node.js](#nodejs)
  - [Browser: Browserify, Webpack, other bundlers](#browser-browserify-webpack-other-bundlers)
  - [Browser: `<script>` Tag](#browser-script-tag)
- [CLI](#cli)
- [API](#api)
  - [Create](#create)
    - [`new PeerId(id[, privKey, pubKey])`](#new-peeridid-privkey-pubkey)
    - [`create([opts])`](#createopts)
  - [Import](#import)
    - [`createFromHexString(str)`](#createfromhexstringstr)
    - [`createFromBytes(buf)`](#createfrombytesbuf)
    - [`createFromCID(cid)`](#createfromcidcid)
    - [`createFromB58String(str)`](#createfromb58stringstr)
    - [`createFromPubKey(pubKey)`](#createfrompubkeypubkey)
    - [`createFromPrivKey(privKey)`](#createfromprivkeyprivkey)
    - [`createFromJSON(obj)`](#createfromjsonobj)
    - [`createFromProtobuf(buf)`](#createfromprotobufbuf)
  - [Export](#export)
    - [`toHexString()`](#tohexstring)
    - [`toBytes()`](#tobytes)
    - [`toString()`](#tostring)
    - [`toB58String()`](#tob58string)
    - [`toJSON()`](#tojson)
    - [`marshal(excludePrivateKey)`](#marshalexcludeprivatekey)
    - [`marshalPubKey()`](#marshalpubkey)
    - [`toPrint()`](#toprint)
    - [`equals(id)`](#equalsid)
    - [`isEqual(id)`](#isequalid)
  - [Others](#others)
    - [`isPeerId(id)`](#ispeeridid)
- [License](#license)

# Description

Generate, import, and export PeerIDs, for use with [IPFS](https://github.com/ipfs/ipfs).

A Peer ID is the SHA-256 [multihash](https://github.com/multiformats/multihash) of a public key.

The public key is a base64 encoded string of a protobuf containing an RSA DER buffer. This uses a node buffer to pass the base64 encoded public key protobuf to the multihash for ID generation.

# Example

```JavaScript
const PeerId = require('peer-id')

const id = await PeerId.create({ bits: 1024, keyType: 'RSA' })
console.log(JSON.stringify(id.toJSON(), null, 2))
```

```bash
{
  "id": "Qma9T5YraSnpRDZqRR4krcSJabThc8nwZuJV3LercPHufi",
  "privKey": "CAAS4AQwggJcAgEAAoGBAMBgbIqyOL26oV3nGPBYrdpbv..",
  "pubKey": "CAASogEwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMBgbIqyOL26oV3nGPBYrdpbvzCY..."
}
```

# Installation

## npm

```sh
> npm i peer-id
```

# Setup

## Node.js

```js
const PeerId = require('peer-id')
```

## Browser: Browserify, Webpack, other bundlers

The code published to npm that gets loaded on require is in fact a ES5
transpiled version with the right shims added. This means that you can require
it and use with your favourite bundler without having to adjust asset management
process.

```js
const PeerId = require('peer-id')
```

## Browser: `<script>` Tag

Loading this module through a script tag will make the `PeerId` obj available in
the global namespace.

```html
<script src="https://unpkg.com/peer-id/dist/index.min.js"></script>
<!-- OR -->
<script src="https://unpkg.com/peer-id/dist/index.js"></script>
```

# CLI

After installing `peer-id`, `npm install peer-id`, you can leverage the cli to generate keys exported as JSON. You can specify the type for the key and size, as detailed in [`create([opts])`](#createopts). The defaults are shown here.

```sh
> peer-id --type rsa --bits 2048
```

# API

```js
const PeerId = require('peer-id')
```

## Create

### `new PeerId(id[, privKey, pubKey])`

- `id: Buffer` - The multihash of the public key as `Buffer`
- `privKey: RsaPrivateKey` - The private key
- `pubKey: RsaPublicKey` - The public key

The key format is detailed in [libp2p-crypto](https://github.com/libp2p/js-libp2p-crypto).

### `create([opts])`

Generates a new Peer ID, complete with public/private keypair.

- `opts.bits: number` - The size of the key. Default: `2048`
- `opts.keyType: string` - The key type, one of: `['RSA', 'Ed25519', 'secp256k1']`. Default: `RSA`

Returns `Promise<PeerId>`.

## Import

### `createFromHexString(str)`

Creates a Peer ID from hex string representing the key's multihash.

Returns `PeerId`.

### `createFromBytes(buf)`

Creates a Peer ID from a buffer representing the key's multihash.

Returns `PeerId`.

### `createFromCID(cid)`

- `cid: CID|String|Buffer` - The multihash encoded as [CID](https://github.com/ipld/js-cid) (object, `String` or `Buffer`)

Creates a Peer ID from a CID representation of the key's multihash ([RFC 0001](https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md)).

Returns `PeerId`.

### `createFromB58String(str)`

Creates a Peer ID from a Base58 string representing the key's multihash.

Returns `PeerId`.

### `createFromPubKey(pubKey)`

- `publicKey: Buffer`

Creates a Peer ID from a buffer containing a public key.

Returns `Promise<PeerId>`.

### `createFromPrivKey(privKey)`

- `privKey: Buffer`

Creates a Peer ID from a buffer containing a private key.

Returns `Promise<PeerId>`.

### `createFromJSON(obj)`

- `obj.id: String` - The multihash encoded in `base58`
- `obj.pubKey: String` - The public key in protobuf format, encoded in `base64`
- `obj.privKey: String` - The private key in protobuf format, encoded in `base64`

Returns `Promise<PeerId>`.

### `createFromProtobuf(buf)`

`buf` is a protocol-buffers encoded PeerId (see `marshal()`)

## Export

### `toHexString()`

Returns the Peer ID's `id` as a hex string.

```
1220d6243998f2fc56343ad7ed0342ab7886a4eb18d736f1b67d44b37fcc81e0f39f
```

### `toBytes()`

Returns the Peer ID's `id` as a buffer.

```
<Buffer 12 20 d6 24 39 98 f2 fc 56 34 3a d7 ed 03 42 ab 78 86 a4 eb 18 d7 36 f1 b6 7d 44 b3 7f cc 81 e0 f3 9f>
```


### `toString()`

Returns the Peer ID's `id` as a self-describing CIDv1 in Base32 ([RFC 0001](https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md))

```
bafzbeigweq4zr4x4ky2dvv7nanbkw6egutvrrvzw6g3h2rftp7gidyhtt4
```

### `toB58String()`

Returns the Peer ID's `id` as a base58 string (multihash/CIDv0).

```
QmckZzdVd72h9QUFuJJpQqhsZqGLwjhh81qSvZ9BhB2FQi
```

### `toJSON()`

Returns an `obj` of the form

- `obj.id: String` - The multihash encoded in `base58`
- `obj.pubKey: String` - The public key in protobuf format, encoded in 'base64'
- `obj.privKey: String` - The private key in protobuf format, encoded in 'base 64'

### `marshal(excludePrivateKey)`

Returns a protocol-buffers encoded version of the id, public key and, if `excludePrivateKey` is not set, the private key.

### `marshalPubKey()`

Returns a protobuf of just the public key, compatible with `libp2p-crypto` (unlike `marshal` above).

For example:
```js
const crypto = require('libp2p-crypto')

PeerId.create({ bits: 256, keyType: 'ed25519' }).then( pid => {
  let pk = crypto.keys.unmarshalPublicKey(pid.marshalPubKey())
  // your code here
}
```

### `toPrint()`

Returns the Peer ID as a printable string without the `Qm` prefix.

Example: `<peer.ID xxxxxx>`

### `equals(id)`

Returns `true` if the given PeerId is equal to the current instance.

- `id` can be a PeerId or a Buffer containing the id

### `isEqual(id)`
**Deprecation Notice**: Use [`equals`](#equalsid), `isEqual` will be removed in 0.14.0.

- `id` can be a PeerId or a Buffer containing the id

## Others

### `isPeerId(id)`

Returns `true` if the given id is an instance of PeerId

- `id` should be an instance of PeerId

# License


MIT


CHANGELOG.md
# [0.16.0](https://github.com/libp2p/js-peer-id/compare/v0.15.4...v0.16.0) (2021-12-01)


### chore

* remove binary ([#160](https://github.com/libp2p/js-peer-id/issues/160)) ([25f0cd4](https://github.com/libp2p/js-peer-id/commit/25f0cd45bda22090994cc48a5085ab79a2be0ace))
* update libp2p-crypto ([#161](https://github.com/libp2p/js-peer-id/issues/161)) ([fbc9bf0](https://github.com/libp2p/js-peer-id/commit/fbc9bf0f9338b41981649659ed32cde92c4c5c28))


### BREAKING CHANGES

* requires node 15+
* removes install cli tool



## [0.15.4](https://github.com/libp2p/js-peer-id/compare/v0.15.3...v0.15.4) (2021-11-26)



## [0.15.3](https://github.com/libp2p/js-peer-id/compare/v0.15.2...v0.15.3) (2021-08-18)



## [0.15.2](https://github.com/libp2p/js-peer-id/compare/v0.15.1...v0.15.2) (2021-07-26)


### Bug Fixes

* use undefined instead of null privKey ([#154](https://github.com/libp2p/js-peer-id/issues/154)) ([35faac2](https://github.com/libp2p/js-peer-id/commit/35faac28f576daf2b7b4d59f8eeacfd0d80cc3af))



## [0.15.1](https://github.com/libp2p/js-peer-id/compare/v0.15.0...v0.15.1) (2021-07-12)


### Bug Fixes

* decode more multibase types ([#152](https://github.com/libp2p/js-peer-id/issues/152)) ([4d74274](https://github.com/libp2p/js-peer-id/commit/4d742746b614f9d0ef29bfb01af7b6ff1fa68cf3))



# [0.15.0](https://github.com/libp2p/js-peer-id/compare/v0.14.8...v0.15.0) (2021-07-06)


### chore

* update to new multiformats ([#149](https://github.com/libp2p/js-peer-id/issues/149)) ([69d7a01](https://github.com/libp2p/js-peer-id/commit/69d7a012cb5efffcc13afe3c75fa1acc8bda1f54))


### BREAKING CHANGES

* uses the CID class from the new multiformats module

Co-authored-by: Vasco Santos <vasco.santos@moxy.studio>



## [0.14.8](https://github.com/libp2p/js-peer-id/compare/v0.14.7...v0.14.8) (2021-04-20)


### Bug Fixes

* specify pbjs root ([#147](https://github.com/libp2p/js-peer-id/issues/147)) ([a125e87](https://github.com/libp2p/js-peer-id/commit/a125e875437f4e36a6a40ba596d819e3f15e519a))



## [0.14.7](https://github.com/libp2p/js-peer-id/compare/v0.14.6...v0.14.7) (2021-04-13)



## [0.14.6](https://github.com/libp2p/js-peer-id/compare/v0.14.4...v0.14.6) (2021-04-06)



## [0.14.4](https://github.com/libp2p/js-peer-id/compare/v0.14.3...v0.14.4) (2021-03-29)


### Bug Fixes

* remove toString side effect that breaks deep equals ([#142](https://github.com/libp2p/js-peer-id/issues/142)) ([eeb5330](https://github.com/libp2p/js-peer-id/commit/eeb53305c8767a6a2b3f610fe9e45a589b4fa972)), closes [#141](https://github.com/libp2p/js-peer-id/issues/141)



## [0.14.3](https://github.com/libp2p/js-peer-id/compare/v0.14.2...v0.14.3) (2021-01-26)



<a name="0.14.2"></a>
## [0.14.2](https://github.com/libp2p/js-peer-id/compare/v0.14.1...v0.14.2) (2020-09-23)


### Features

* has inline public key method ([#132](https://github.com/libp2p/js-peer-id/issues/132)) ([b2ee342](https://github.com/libp2p/js-peer-id/commit/b2ee342))



<a name="0.14.1"></a>
## [0.14.1](https://github.com/libp2p/js-peer-id/compare/v0.13.13...v0.14.1) (2020-09-03)


### Bug Fixes

* privKey possible undefined ([#129](https://github.com/libp2p/js-peer-id/issues/129)) ([224b30c](https://github.com/libp2p/js-peer-id/commit/224b30c))
* replace node buffers with uint8arrays ([#127](https://github.com/libp2p/js-peer-id/issues/127)) ([d16ce9c](https://github.com/libp2p/js-peer-id/commit/d16ce9c))
* ts constructor types ([#130](https://github.com/libp2p/js-peer-id/issues/130)) ([d40d588](https://github.com/libp2p/js-peer-id/commit/d40d588))
* typo in readme ([#128](https://github.com/libp2p/js-peer-id/issues/128)) ([6d571ae](https://github.com/libp2p/js-peer-id/commit/6d571ae))


### BREAKING CHANGES

* - Where node Buffers were returned, now Uint8Arrays are

* chore: remove gh dep



<a name="0.14.0"></a>
# [0.14.0](https://github.com/libp2p/js-peer-id/compare/v0.13.13...v0.14.0) (2020-08-07)


### Bug Fixes

* replace node buffers with uint8arrays ([#127](https://github.com/libp2p/js-peer-id/issues/127)) ([d16ce9c](https://github.com/libp2p/js-peer-id/commit/d16ce9c))


### BREAKING CHANGES

* - Where node Buffers were returned, now Uint8Arrays are

* chore: remove gh dep



<a name="0.13.13"></a>
## [0.13.13](https://github.com/libp2p/js-peer-id/compare/v0.13.12...v0.13.13) (2020-06-23)



<a name="0.13.12"></a>
## [0.13.12](https://github.com/libp2p/js-peer-id/compare/v0.13.11...v0.13.12) (2020-04-22)


### Features

* **cli:** add support for specifying type and size ([#122](https://github.com/libp2p/js-peer-id/issues/122)) ([8cd9dfb](https://github.com/libp2p/js-peer-id/commit/8cd9dfb))



<a name="0.13.11"></a>
## [0.13.11](https://github.com/libp2p/js-peer-id/compare/v0.13.10...v0.13.11) (2020-03-26)



<a name="0.13.10"></a>
## [0.13.10](https://github.com/libp2p/js-peer-id/compare/v0.13.9...v0.13.10) (2020-03-18)


### Bug Fixes

* add buffer ([#120](https://github.com/libp2p/js-peer-id/issues/120)) ([c305c36](https://github.com/libp2p/js-peer-id/commit/c305c36))



<a name="0.13.9"></a>
## [0.13.9](https://github.com/libp2p/js-peer-id/compare/v0.13.8...v0.13.9) (2020-02-19)



<a name="0.13.8"></a>
## [0.13.8](https://github.com/libp2p/js-peer-id/compare/v0.13.6...v0.13.8) (2020-02-18)


### Bug Fixes

* remove use of assert module ([#117](https://github.com/libp2p/js-peer-id/issues/117)) ([f44645e](https://github.com/libp2p/js-peer-id/commit/f44645e))


### Features

* adds typescript types + type tests ([#110](https://github.com/libp2p/js-peer-id/issues/110)) ([a5070ae](https://github.com/libp2p/js-peer-id/commit/a5070ae))




karma.conf.js
module.exports = (config) => {
  const path = require('path')
  const node_modules_dir = path.join(__dirname, 'node_modules')
  const deps = [
    'deps/forge.bundle.js'
  ]
  config.set({
    basePath: '',
    frameworks: ['mocha'],

    files: [
      'tests/test.js'
    ],

    preprocessors: {
      'tests/*': ['webpack']
    },

    webpack: {
      output: {
        path: path.join(__dirname, 'dist'),
        filename: 'bundle.js'
      },
      resolve: {
        extensions: ['', '.js', '.json'],
        alias: { 'node-forge': path.resolve(__dirname, 'deps/forge.bundle.js') }
      },
      externals: {
        fs: '{}'
      },
      node: {
        Buffer: true
      },
      module: {
        loaders: [
          { test: /\.json$/, loader: 'json' }
        ],
        noParse: []
      }
    },

    webpackMiddleware: {
      noInfo: true,
      stats: {
        colors: true
      }
    },
    reporters: ['spec'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: false,
    browsers: process.env.TRAVIS ? ['Firefox'] : ['Chrome'],
    singleRun: true
  })

  deps.forEach((dep) => {
    const depPath = path.resolve(node_modules_dir, dep)
    config.webpack.module.noParse.push(depPath)
  })
}


src/index.js
/*
 * Id is an object representation of a peer Id. a peer Id is a multihash
 */

'use strict'

const { CID } = require('multiformats/cid')
const b32 = require('multiformats/bases/base32')
const b36 = require('multiformats/bases/base36')
const b58 = require('multiformats/bases/base58')
const b64 = require('multiformats/bases/base64')
const { base58btc } = require('multiformats/bases/base58')
const { base32 } = require('multiformats/bases/base32')
const { base16 } = require('multiformats/bases/base16')
const Digest = require('multiformats/hashes/digest')
const cryptoKeys = require('libp2p-crypto/src/keys')
const withIs = require('class-is')
const { PeerIdProto } = require('./proto')
const uint8ArrayEquals = require('uint8arrays/equals')
const uint8ArrayFromString = require('uint8arrays/from-string')
const uint8ArrayToString = require('uint8arrays/to-string')
const { identity } = require('multiformats/hashes/identity')

const bases = {
  ...b32,
  ...b36,
  ...b58,
  ...b64
}
const baseDecoder = Object.keys(bases).reduce(
  (acc, curr) => acc.or(bases[curr]),
  base32.decoder
)

// these values are from https://github.com/multiformats/multicodec/blob/master/table.csv
const DAG_PB_CODE = 0x70
const LIBP2P_KEY_CODE = 0x72

class PeerId {
  constructor (id, privKey, pubKey) {
    if (!(id instanceof Uint8Array)) {
      throw new Error('invalid id provided')
    }

    if (privKey && pubKey && !uint8ArrayEquals(privKey.public.bytes, pubKey.bytes)) {
      throw new Error('inconsistent arguments')
    }

    this._id = id
    this._idB58String = base58btc.encode(this.id).substring(1)
    this._privKey = privKey
    this._pubKey = pubKey
  }

  get id () {
    return this._id
  }

  set id (val) {
    throw new Error('Id is immutable')
  }

  get privKey () {
    return this._privKey
  }

  set privKey (privKey) {
    this._privKey = privKey
  }

  get pubKey () {
    if (this._pubKey) {
      return this._pubKey
    }

    if (this._privKey) {
      return this._privKey.public
    }

    try {
      const decoded = Digest.decode(this.id)

      if (decoded.code === identity.code) {
        this._pubKey = cryptoKeys.unmarshalPublicKey(decoded.digest)
      }
    } catch (_) {
      // Ignore, there is no valid public key
    }

    return this._pubKey
  }

  set pubKey (pubKey) {
    this._pubKey = pubKey
  }

  // Return the protobuf version of the public key, matching go ipfs formatting
  marshalPubKey () {
    if (this.pubKey) {
      return cryptoKeys.marshalPublicKey(this.pubKey)
    }
  }

  // Return the protobuf version of the private key, matching go ipfs formatting
  marshalPrivKey () {
    if (this.privKey) {
      return cryptoKeys.marshalPrivateKey(this.privKey)
    }
  }

  // Return the protobuf version of the peer-id
  marshal (excludePriv) {
    return PeerIdProto.encode({
      id: this.toBytes(),
      pubKey: this.marshalPubKey(),
      privKey: excludePriv ? null : this.marshalPrivKey()
    }).finish()
  }

  toPrint () {
    let pid = this.toB58String()
    // All sha256 nodes start with Qm
    // We can skip the Qm to make the peer.ID more useful
    if (pid.startsWith('Qm')) {
      pid = pid.slice(2)
    }
    let maxRunes = 6
    if (pid.length < maxRunes) {
      maxRunes = pid.length
    }

    return '<peer.ID ' + pid.substr(0, maxRunes) + '>'
  }

  // return the jsonified version of the key, matching the formatting
  // of go-ipfs for its config file
  toJSON () {
    return {
      id: this.toB58String(),
      privKey: toB64Opt(this.marshalPrivKey()),
      pubKey: toB64Opt(this.marshalPubKey())
    }
  }

  // encode/decode functions
  toHexString () {
    return base16.encode(this.id).substring(1)
  }

  toBytes () {
    return this.id
  }

  toB58String () {
    return this._idB58String
  }

  // return self-describing String representation
  // in default format from RFC 0001: https://github.com/libp2p/specs/pull/209
  toString () {
    if (!this._idCIDString) {
      const cid = CID.createV1(LIBP2P_KEY_CODE, Digest.decode(this.id))

      Object.defineProperty(this, '_idCIDString', {
        value: cid.toString(),
        enumerable: false
      })
    }
    return this._idCIDString
  }

  /**
   * Checks the equality of `this` peer against a given PeerId.
   *
   * @param {Uint8Array|PeerId} id
   * @returns {boolean}
   */
  equals (id) {
    if (id instanceof Uint8Array) {
      return uint8ArrayEquals(this.id, id)
    } else if (id.id) {
      return uint8ArrayEquals(this.id, id.id)
    } else {
      throw new Error('not valid Id')
    }
  }

  /**
   * Checks the equality of `this` peer against a given PeerId.
   *
   * @deprecated Use `.equals`
   * @param {Uint8Array|PeerId} id
   * @returns {boolean}
   */
  isEqual (id) {
    return this.equals(id)
  }

  /*
   * Check if this PeerId instance is valid (privKey -> pubKey -> Id)
   */
  isValid () {
    // TODO: needs better checking
    return Boolean(this.privKey &&
      this.privKey.public &&
      this.privKey.public.bytes &&
      this.pubKey.bytes instanceof Uint8Array &&
        uint8ArrayEquals(this.privKey.public.bytes, this.pubKey.bytes))
  }

  /**
   * Check if the PeerId has an inline public key.
   *
   * @returns {boolean}
   */
  hasInlinePublicKey () {
    try {
      const decoded = Digest.decode(this.id)

      if (decoded.code === identity.code) {
        return true
      }
    } catch (_) {
      // Ignore, there is no valid public key
    }

    return false
  }
}

const PeerIdWithIs = withIs(PeerId, {
  className: 'PeerId',
  symbolName: '@libp2p/js-peer-id/PeerId'
})

exports = module.exports = PeerIdWithIs

const computeDigest = (pubKey) => {
  if (pubKey.bytes.length <= 42) {
    return Digest.create(identity.code, pubKey.bytes).bytes
  } else {
    return pubKey.hash()
  }
}

const computePeerId = async (privKey, pubKey) => {
  const digest = await computeDigest(pubKey)
  return new PeerIdWithIs(digest, privKey, pubKey)
}

// generation
exports.create = async (opts) => {
  opts = opts || {}
  opts.bits = opts.bits || 2048
  opts.keyType = opts.keyType || 'RSA'

  const key = await cryptoKeys.generateKeyPair(opts.keyType, opts.bits)
  return computePeerId(key, key.public)
}

exports.createFromHexString = (str) => {
  return new PeerIdWithIs(base16.decode('f' + str))
}

exports.createFromBytes = (buf) => {
  try {
    const cid = CID.decode(buf)

    if (!validMulticodec(cid)) {
      throw new Error('Supplied PeerID CID is invalid')
    }

    return exports.createFromCID(cid)
  } catch {
    const digest = Digest.decode(buf)

    if (digest.code !== identity.code) {
      throw new Error('Supplied PeerID CID is invalid')
    }

    return new PeerIdWithIs(buf)
  }
}

exports.createFromB58String = (str) => {
  return exports.createFromBytes(base58btc.decode('z' + str))
}

const validMulticodec = (cid) => {
  // supported: 'libp2p-key' (CIDv1) and 'dag-pb' (CIDv0 converted to CIDv1)
  return cid.code === LIBP2P_KEY_CODE || cid.code === DAG_PB_CODE
}

exports.createFromCID = (cid) => {
  cid = CID.asCID(cid)

  if (!cid || !validMulticodec(cid)) {
    throw new Error('Supplied PeerID CID is invalid')
  }

  return new PeerIdWithIs(cid.multihash.bytes)
}

// Public Key input will be a Uint8Array
exports.createFromPubKey = async (key) => {
  let buf = key

  if (typeof buf === 'string') {
    buf = uint8ArrayFromString(key, 'base64pad')
  }

  if (!(buf instanceof Uint8Array)) {
    throw new Error('Supplied key is neither a base64 string nor a Uint8Array')
  }

  const pubKey = await cryptoKeys.unmarshalPublicKey(buf)
  return computePeerId(undefined, pubKey)
}

// Private key input will be a string
exports.createFromPrivKey = async (key) => {
  if (typeof key === 'string') {
    key = uint8ArrayFromString(key, 'base64pad')
  }

  if (!(key instanceof Uint8Array)) {
    throw new Error('Supplied key is neither a base64 string nor a Uint8Array')
  }

  const privKey = await cryptoKeys.unmarshalPrivateKey(key)
  return computePeerId(privKey, privKey.public)
}

exports.createFromJSON = async (obj) => {
  const id = base58btc.decode('z' + obj.id)
  const rawPrivKey = obj.privKey && uint8ArrayFromString(obj.privKey, 'base64pad')
  const rawPubKey = obj.pubKey && uint8ArrayFromString(obj.pubKey, 'base64pad')
  const pub = rawPubKey && await cryptoKeys.unmarshalPublicKey(rawPubKey)

  if (!rawPrivKey) {
    return new PeerIdWithIs(id, undefined, pub)
  }

  const privKey = await cryptoKeys.unmarshalPrivateKey(rawPrivKey)
  const privDigest = await computeDigest(privKey.public)

  let pubDigest

  if (pub) {
    pubDigest = await computeDigest(pub)
  }

  if (pub && !uint8ArrayEquals(privDigest, pubDigest)) {
    throw new Error('Public and private key do not match')
  }

  if (id && !uint8ArrayEquals(privDigest, id)) {
    throw new Error('Id and private key do not match')
  }

  return new PeerIdWithIs(id, privKey, pub)
}

exports.createFromProtobuf = async (buf) => {
  if (typeof buf === 'string') {
    buf = uint8ArrayFromString(buf, 'base16')
  }

  let { id, privKey, pubKey } = PeerIdProto.decode(buf)

  privKey = privKey ? await cryptoKeys.unmarshalPrivateKey(privKey) : false
  pubKey = pubKey ? await cryptoKeys.unmarshalPublicKey(pubKey) : false

  let pubDigest
  let privDigest

  if (privKey) {
    privDigest = await computeDigest(privKey.public)
  }

  if (pubKey) {
    pubDigest = await computeDigest(pubKey)
  }

  if (privKey) {
    if (pubKey) {
      if (!uint8ArrayEquals(privDigest, pubDigest)) {
        throw new Error('Public and private key do not match')
      }
    }
    return new PeerIdWithIs(privDigest, privKey, privKey.public)
  }

  // TODO: val id and pubDigest

  if (pubKey) {
    return new PeerIdWithIs(pubDigest, undefined, pubKey)
  }

  if (id) {
    return new PeerIdWithIs(id)
  }

  throw new Error('Protobuf did not contain any usable key material')
}

exports.parse = (str) => {
  if (str.charAt(0) === '1' || str.charAt(0) === 'Q') {
    // identity hash ed25519 key or sha2-256 hash of rsa public key
    // base58btc encoded either way
    str = `z${str}`
  }

  return exports.createFromBytes(baseDecoder.decode(str))
}

exports.isPeerId = (peerId) => {
  return Boolean(typeof peerId === 'object' &&
    peerId._id &&
    peerId._idB58String)
}

function toB64Opt (val) {
  if (val) {
    return uint8ArrayToString(val, 'base64pad')
  }
}
